File: DR/1_27/2110.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
	int n, c, max_dist = 0;
	cin>>n>>c;
	
	vector<int> v(n);
	for(int i = 0; i<n; i++) {
		cin>>v[i];
	}
	
	sort(v.begin(), v.end());
	
	int start = 0, end = v[n-1], ip_cnt;
	while(start<=end) {
		int mid = (start + end) / 2, ip_cnt = 1;
		int prev = v[0];
		for(int i = 1; i<n; i++) { 
			if(v[i] - prev >= mid) {
				prev = v[i];
				ip_cnt++;
			} 
		}
		if(ip_cnt >= c) {
			start = mid + 1;
			max_dist = max(max_dist, mid);
		} else {
			end = mid - 1;
		}
	}
	cout<<max_dist;
}

File: MJ/2.13/binarysearch/2110.py
# 7:00 -> 7:40 + 0:20 -> 8:00
# GG
import sys
sys_input = sys.stdin.readline

N, C = map(int, input().split())


max_distance = 0
homes = []
for _ in range(N):
    homes.append(int(sys_input()))
homes.sort()
full_distance = homes[len(homes)-1]-homes[0]


def binary_search(start, end, homes):
    while start <= end:
        mid = (start+end)//2
        cnt_wifi = 1
        pre_home = homes[0]
        for i in range(1, len(homes)):
            if homes[i] >= mid+pre_home:
                cnt_wifi += 1
                pre_home = homes[i]
        if cnt_wifi >= C:
            global max_distance
            max_distance = max(max_distance, mid)
            start = mid+1
        else:
            end = mid-1


binary_search(1, full_distance, homes)
print(max_distance)
